README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# treeco <img src="man/figures/logo.png" align="right" height=150/>

[![Travis build
status](https://travis-ci.org/tyluRp/treeco.svg?branch=master)](https://travis-ci.org/tyluRp/treeco)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/tyluRp/treeco?branch=master&svg=true)](https://ci.appveyor.com/project/tyluRp/treeco)
[![Coverage
status](https://codecov.io/gh/tyluRp/treeco/branch/master/graph/badge.svg)](https://codecov.io/github/tyluRp/treeco?branch=master)
[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)

<img src="man/figures/co2_plot.png" align="center"/>

The goal of `treeco` is to provide R users a tool for calculating the
eco benefits of trees. All data used to calculate benefits is ripped
from OpenStreetMaps
[`otm-ecoservice`](https://github.com/OpenTreeMap/otm-ecoservice)
repository which was (probably) ripped from
[i-Tree](https://www.itreetools.org/)’s Eco or Streets software. A
single tree is represented by 15 rows and 8 columns as there are 15
benefits calculated for every tree. Since tree inventories can be rather
large, `treeco` utilizes the
[`data.table`](https://github.com/Rdatatable/data.table) package for
speed. All calculations are done on unique species/dbh pairs to avoid
redundant computation.

## Installation

`treeco` isn’t available on CRAN but you can install it directly from
github using [`devtools`](https://github.com/r-lib/devtools):

``` r
# install.packages("devtools")
devtools::install_github("tylurp/treeco")
```

## A reproducible example

We can use the
[`trees`](https://stat.ethz.ch/R-manual/R-patched/library/datasets/html/trees.html)
dataset to demonstrate how `eco_guess` and `eco_run_all` works:

``` r
library(dplyr)
library(treeco)

df_trees <- trees %>% 
  mutate(common_name = "Black cherry") %>% 
  select(common_name, Girth) %>% 
  mutate(botanical_name = eco_guess(common_name, "botanical"))

eco_run_all(
  data = df_trees,                  # dataset or path to CSV
  common_col = "common_name",       # common name field
  botanical_col = "botanical_name", # botanical name field
  dbh_col = "Girth",                # dbh field
  region = "PiedmtCLT",             # region code
  n = 0.99,                         # optional, threshold for species guessing
  ) %>% as_tibble()
#> # A tibble: 465 x 8
#>    botanical    common     dbh benefit_value benefit   unit  dollars rn   
#>    <chr>        <chr>    <dbl>         <dbl> <chr>     <chr>   <dbl> <chr>
#>  1 prunus sero… black c…   8.3        0.0776 aq nox a… lb       0.51 1    
#>  2 prunus sero… black c…   8.3        0.026  aq nox d… lb       0.17 1    
#>  3 prunus sero… black c…   8.3        0.0556 aq ozone… lb       0.36 1    
#>  4 prunus sero… black c…   8.3        0.015  aq pm10 … lb       0.04 1    
#>  5 prunus sero… black c…   8.3        0.0633 aq pm10 … lb       0.16 1    
#>  6 prunus sero… black c…   8.3        0.165  aq sox a… lb       0.32 1    
#>  7 prunus sero… black c…   8.3        0.0119 aq sox d… lb       0.02 1    
#>  8 prunus sero… black c…   8.3        0.0146 aq voc a… lb       0.09 1    
#>  9 prunus sero… black c…   8.3       -0.338  bvoc      lb       2.12 1    
#> 10 prunus sero… black c…   8.3       30.3    co2 avoi… lb       0.23 1    
#> # ... with 455 more rows
```

## More examples

Use `eco_run` to calculate benefits for a single tree:

``` r
treeco::eco_run("Common fig", 20, "InlEmpCLM")
#>     common_name dbh benefit_value            benefit unit dollars
#>  1:  Common fig  20        0.2429     aq nox avoided   lb    0.93
#>  2:  Common fig  20        0.2623         aq nox dep   lb    1.01
#>  3:  Common fig  20        0.7716       aq ozone dep   lb    2.96
#>  4:  Common fig  20        0.0602    aq pm10 avoided   lb    0.28
#>  5:  Common fig  20        0.4079        aq pm10 dep   lb    1.89
#>  6:  Common fig  20        0.4813     aq sox avoided   lb    1.17
#>  7:  Common fig  20        0.0353         aq sox dep   lb    0.09
#>  8:  Common fig  20        0.0602     aq voc avoided   lb    0.12
#>  9:  Common fig  20        0.0000               bvoc   lb    0.00
#> 10:  Common fig  20      122.7973        co2 avoided   lb    0.41
#> 11:  Common fig  20        9.0389    co2 sequestered   lb    0.03
#> 12:  Common fig  20     1255.7516        co2 storage   lb    4.19
#> 13:  Common fig  20      189.2000        electricity  kwh   38.12
#> 14:  Common fig  20      834.7837 hydro interception  gal    4.59
#> 15:  Common fig  20     -179.4561        natural gas   lb    1.20
```

One issue with eco benefits is that they all rely on i-Tree’s
`master_species_list` which is a list of 3,000+ species, therefore a
users data needs to fit this list in order to extract benefits. For
example, “Commn fig” doesn’t match i-Tree’s “Common fig” because of the
typo. So far, there really isn’t a great solution to this. For now,
`treeco` guesses the species code on the fly by quantifying the
“similarity”, anything below 90% similar is immediately discarded.

For example, if we misspell “Common fig” as “Commn fig”:

``` r
treeco::eco_run("Commn fig", 20, "InlEmpCLM")
#> Species given: [commn fig]
#> Closest match: [common fig]
#> ...
#> Using closest match
#>     common_name dbh benefit_value            benefit unit dollars
#>  1:  Common fig  20        0.2429     aq nox avoided   lb    0.93
#>  2:  Common fig  20        0.2623         aq nox dep   lb    1.01
#>  3:  Common fig  20        0.7716       aq ozone dep   lb    2.96
#>  4:  Common fig  20        0.0602    aq pm10 avoided   lb    0.28
#>  5:  Common fig  20        0.4079        aq pm10 dep   lb    1.89
#>  6:  Common fig  20        0.4813     aq sox avoided   lb    1.17
#>  7:  Common fig  20        0.0353         aq sox dep   lb    0.09
#>  8:  Common fig  20        0.0602     aq voc avoided   lb    0.12
#>  9:  Common fig  20        0.0000               bvoc   lb    0.00
#> 10:  Common fig  20      122.7973        co2 avoided   lb    0.41
#> 11:  Common fig  20        9.0389    co2 sequestered   lb    0.03
#> 12:  Common fig  20     1255.7516        co2 storage   lb    4.19
#> 13:  Common fig  20      189.2000        electricity  kwh   38.12
#> 14:  Common fig  20      834.7837 hydro interception  gal    4.59
#> 15:  Common fig  20     -179.4561        natural gas   lb    1.20
```

If you are missing a field, you can use `eco_guess` to try and find it:

``` r
x <- c("common fig", "red maple", "fir")
treeco::eco_guess(x, "botanical")
#> [1] "ficus carica" "acer rubrum"  "abies spp"
```


NEWS.md
# treeco 0.1.0.9000
* All datasets have been restructured and are more 'tidy'. Old datasets were the same as sysdata. Also, if users have the `tibble` package loaded, the datasets will return tibbles.

# treeco 0.0.1.9000

* Added `pkgdown` support.
* Fixes bug where `eco_guess` only guessed unique values of `x`, this meant a 
user with duplicates would have issues assigning `x` to a `data.frame` as their lengths would differ.
* Added threshold for guessing in `eco_run_all`, 0.8 is the default value.
* Converted `message` to `warning` in `eco_guess` when `x` vector is coerced from `factor` to `character`.
* Added warnings to `eco_run_all` for cases where `n` > 1 or `n` < 0. If this happens, `n` defaults to 0.8.
* Added tests for warnings where threshold doesn't make sense and where factors converted to character.
* Modified `eco_guess` to make it faster. This function takes a vector, converts it to a `data.table`, and does all the matching on unique values only. Then joins the missing field to the original (non unique) `data.table` and returns a vector containing values for the missing field.
* The utility function `extract_data` now preserves the original rownames as a column `rn` using the `keep.rownames = TRUE` argument in `as.data.table`. This allows users to join additional variables back to the benefits dataframe. Especially important given this type of data is usually spatial. There is still the issue of reading data as the `fread` function isn't preserving rownames. Will come back to that at some point. 
* Added [`Vignettes`](http://r-pkgs.had.co.nz/vignettes.html).
* Added first vignette, getting started article.
* Using RMarkdown to generate README.md
* Added unit parameter for rare cases where user data measures DBH in centimeters instead of inches.

# treeco 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/eco_run.R
#' Eco benefits calculator
#'
#' @description Run the eco benefits for a single tree. This function acts like
#' a benefits calculator in that you give a common name and it calculates the
#' benefits. Similar to Davey Tree's benefits calculator.
#'
#' @param common the common name of the tree
#' @param dbh the dbh value in inches
#' @param region region code, see \code{money} dataset for region codes
#'
#' @examples
#' eco_run("common fig", 20, "InlEmpCLM")
#'
#' eco_run("red maple", 15, "PacfNWLOG")
#'
#' @importFrom  data.table as.data.table melt setnames setkey
#' @export
eco_run <- function(common, dbh, region) {

  # To avoid notes about global variables
  # See: https://github.com/Rdatatable/data.table/issues/850
  benefit_value = NULL
  unit = NULL
  benefit = NULL
  .SD = NULL
  `:=` = NULL

  species_guess <- guess_common(common, region)
  species_val <- species_guess$spp_value_assignment
  species_common_guess <- species_guess$common_name

  tree     <- data.table(common_name = species_common_guess,
                                     dbh_val = dbh,
                                     region = region,
                                     stringsAsFactors = FALSE)
  benefits <- as.data.table(benefits)
  species  <- as.data.table(species)
  money    <- as.data.table(money)
  money    <- money[money$region_code == region, ]                               # Filter currency dataset by region
  money    <- melt(money, id.vars = c("region_code", "region_name")) # Melt the dataset to 'tidy' format
  money    <- money[, c("variable", "value")]                                    # Select the variables we need

  benefits <- benefits[benefits$species_region == region]                        # Extract benefits within user defined region
  species <- species[species$species_region == region]                            # Extract species within user defined region

  tree$dbh_val <- tree$dbh_val * 2.54

  tree$spp_value <- species_val

  benefits <- benefits[benefits$species_code == species_val]

  setkey(tree, "spp_value")
  setkey(benefits, "species_code")

  tree <- tree[benefits, allow.cartesian=TRUE]

  dbh_values <- tree[["dbh_val"]]
  dbh_ranges <- c(3.81, 11.43, 22.86, 38.10, 53.34, 68.58, 83.82, 99.06, 114.30)

  z <- abs(outer(dbh_values, dbh_ranges, `-`))

  # A function that returns the position of n-th largest
  # https://stackoverflow.com/questions/10296866/
  tree$a <- apply(z, 1, which.min)
  z[z==apply(z, 1, min)] <- Inf
  tree$b <- apply(z, 1, which.min)
  tree$start <- pmin(tree$a, tree$b)
  tree$end <- pmax(tree$a, tree$b)
  tree$x1 <- dbh_ranges[tree$start]
  tree$x2 <- dbh_ranges[tree$end]

  # Grab the benefit values given the start/end vectors
  # https://stackoverflow.com/questions/20617371/
  tbl_rows <- seq_along(tree$id)
  tbl_indicies_y1 <- tree$start
  tbl_mat <- as.matrix(tree[,6:14]) # careful with this
  tree$y1 <- tbl_mat[cbind(tbl_rows, tbl_indicies_y1)]

  tbl_rows <- seq_along(tree$id)
  tbl_indicies_y2 <- tree$end
  tbl_mat <- as.matrix(tree[,6:14]) # careful with this, easy to fuck up
  tree$y2 <- tbl_mat[cbind(tbl_rows, tbl_indicies_y2)]

  tree[, "benefit_value" := ifelse(tree$y1 == tree$y2, tree$y1, eco_interp(x = tree$dbh_val, x1 = tree$x1, x2 = tree$x2, y1 = tree$y1, y2 = tree$y2))]
  tree$benefit_value <- round(tree$benefit_value, 4)

  # A bunch of copy/paste (yuck) to grab eco benefit money values so we can
  # multiply the benefit by these values to get $ saved. Figure out a more
  # elegant way of doing this. Looks ugly.
  elec_money <- money[grepl("electricity", money$variable)][["value"]]
  gas_money  <- money[grepl("natural_gas", money$variable)][["value"]]
  h20_money  <- money[grepl("h20_gal", money$variable)][["value"]]
  co2_money  <- money[grepl("co2", money$variable)][["value"]]
  o3_money   <- money[grepl("o3_lb", money$variable)][["value"]]
  nox_money  <- money[grepl("nox_lb", money$variable)][["value"]]
  pm10_money <- money[grepl("pm10_lb", money$variable)][["value"]]
  sox_money  <- money[grepl("sox_lb", money$variable)][["value"]]
  voc_money  <- money[grepl("voc_lb", money$variable)][["value"]]

  # Convert kilograms and cubic meters to lbs and gallons
  tree[grepl("lb", unit), "benefit_value" := benefit_value * 2.20462]
  tree[grepl("gal", unit), "benefit_value" := benefit_value * 264.172052]

  # Multiply $ values by the benefit value to get the $ saved
  tree[grepl("electricity", benefit), "dollars" := benefit_value * elec_money]
  tree[grepl("natural gas", benefit), "dollars" := benefit_value * gas_money]
  tree[grepl("hydro interception", benefit), "dollars" := benefit_value * h20_money]
  tree[grepl("co2 ", benefit), "dollars" := benefit_value * co2_money]
  tree[grepl("aq ozone dep", benefit), "dollars" := benefit_value * o3_money]
  tree[grepl("aq nox", benefit), "dollars" := benefit_value * nox_money]
  tree[grepl("aq pm10", benefit), "dollars" := benefit_value * pm10_money]
  tree[grepl("aq sox", benefit), "dollars" := benefit_value * sox_money]
  tree[grepl("voc", benefit), "dollars" := benefit_value * voc_money]

  # Because davey takes $ values like -0.54 from natural gas and makes it 0.54?
  tree$dollars <- abs(round(tree$dollars, 2))
  tree$benefit_value <- round(tree$benefit_value, 4)

  setkey(tree, NULL)

  tree$dbh_val <- round(tree$dbh_val * 0.393701, 2)

  tree_vars <- c("common_name", "dbh_val", "benefit_value", "benefit", "unit", "dollars")
  tree <- tree[, .SD, .SDcols = tree_vars]
  setnames(tree, "dbh_val", "dbh")
  tree$common_name <- paste0(toupper(substr(tree$common_name, 1, 1)), substr(tree$common_name, 2, nchar(tree$common_name)))

  return(tree)

}

